subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1598,15 +1688,72 @@ public Configuration attributes(Action<? super AttributeContainer> action) {
         return this;
     }
 
+    @Override
+    public void preventUsageMutation() {
+        usageCanBeMutated = false;
+    }
+
+    public boolean isUsageMutable() {
+        return usageCanBeMutated;
+    }
+
+    @SuppressWarnings("deprecation")
+    private void assertUsageIsMutable() {
+        if (!usageCanBeMutated) {
+            // Don't print role message for legacy role - users might not have actively chosen this role
+            if (roleAtCreation != null && roleAtCreation != ConfigurationRoles.LEGACY) {
+                throw new GradleException(
+                        String.format("Cannot change the allowed usage of %s, as it was locked upon creation to the role: '%s'.\n" +
+                                "This role permits the following usage:\n" +
+                                "%s\n" +
+                                "Ideally, each configuration should be used for a single purpose.",
+                                getDisplayName(), roleAtCreation.getName(), roleAtCreation.describe()));
+            } else {
+                throw new GradleException(String.format("Cannot change the allowed usage of %s, as it has been locked.", getDisplayName()));
+            }
+        }
+    }
+
+    private void logChangingUsage(String usage, boolean allowed) {
+        String msgTemplate = "Configuration {} allowed usage is changing: {}. Ideally, usage should be fixed upon creation.";
+        if (warnOnChangingUsage) {
+            LOGGER.warn(msgTemplate, getDisplayName(), describeChangingUsage(usage, allowed));
```

Comment:
I think we'll want a full deprecation warning here. Though, that can wait for 8.1. Might be worth adding a TODO here to change this to a `DeprecationLogger`

Reply:
For 8.1, definitely Deprecation Logger.  I've added a note to the field about removing.  I guess, since this is opt-in, there's no harm in using dep logger now.  I'll update it.